Lib/socketserver.py

```diff
@@ -651,24 +652,32 @@ def process_request_thread(self, request, client_address):
         except Exception:
             self.handle_error(request, client_address)
         finally:
-            self.shutdown_request(request)
+            try:
+                self.shutdown_request(request)
+            finally:
+                thread = threading.current_thread()
+                with self._threads_lock:
+                    if self._threads and not thread.daemon:
+                        self._threads.remove(thread)
 
     def process_request(self, request, client_address):
         """Start a new thread to process the request."""
         t = threading.Thread(target = self.process_request_thread,
                              args = (request, client_address))
         t.daemon = self.daemon_threads
         if not t.daemon and self.block_on_close:
-            if self._threads is None:
-                self._threads = []
-            self._threads.append(t)
+            with self._threads_lock:
+                if self._threads is None:
+                    self._threads = []
+                self._threads.append(t)
         t.start()
 
     def server_close(self):
         super().server_close()
         if self.block_on_close:
             threads = self._threads
-            self._threads = None
+            with self._threads_lock:
```

Comment:
You must move "threads = self._threads" inside the with block for correctness.

Reply:
Creation of `self._threads` is now done with `set_default` to make it as atomic as possible without a lock outside the instance.